build-plugins/build-support/src/main/kotlin/aws/sdk/kotlin/gradle/plugins/artifactmetrics/ArtifactMetricsPlugin.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.sdk.kotlin.gradle.plugins.artifactmetrics
+
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.tasks.TaskProvider
+import org.gradle.jvm.tasks.Jar
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.withType
+
+private const val TASK_GROUP = "Metrics"
+internal const val OUTPUT_PATH = "reports/metrics/"
+
+/**
+ * Facilitates the collection and analysis of artifact size metrics via the `artifactMetrics` and `analyzeArtifactMetrics` gradle tasks.
+ */
+class ArtifactMetricsPlugin : Plugin<Project> {
+    override fun apply(target: Project) {
+        if (target != target.rootProject) {
+            throw GradleException("${this::class.java} can only be applied to the root project")
+        }
+
+        val tasks = mutableListOf<TaskProvider<CollectMetricsTask>>()
+        target.subprojects {
+            if (
+                (target.rootProject.name == "aws-sdk-kotlin" && (path.startsWith(":services") || path.startsWith(":aws-runtime"))) ||
+                (target.rootProject.name == "smithy-kotlin" && (path.startsWith(":codegen") || path.startsWith(":runtime"))) ||
+                (target.rootProject.name == "aws-crt-kotlin-parent" && path.startsWith(":aws-crt-kotlin"))
+            ) {
+                logger.info("registering artifact metrics tasks for $name at $path")
+                tasks.add(subprojectMetricsTask())
+            }
```

Comment:
**Correctness**: The determination of which subprojects to include/exclude should be made inside the consuming projects as plugin/task config. We don't want to have to make changes in the tools repo whenever we make new projects in the consuming repos.